src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.State.cs

```diff
@@ -54,14 +54,22 @@ private class State
                 /// </summary>
                 public virtual ValueSource<Optional<Compilation>>? FinalCompilation => null;
 
-                protected State(ValueSource<Optional<Compilation>>? compilation, Compilation? declarationOnlyCompilation, TrackedGeneratorDriver generatorDriver)
+                protected State(
+                    ValueSource<Optional<Compilation>>? compilation,
+                    Compilation? declarationOnlyCompilation,
+                    TrackedGeneratorDriver generatorDriver,
+                    Compilation? compilationDoNotKeep)
                 {
                     // Declaration-only compilations should never have any references
                     Contract.ThrowIfTrue(declarationOnlyCompilation != null && declarationOnlyCompilation.ExternalReferences.Any());
 
                     Compilation = compilation;
                     DeclarationOnlyCompilation = declarationOnlyCompilation;
                     GeneratorDriver = generatorDriver;
+
+                    if (compilationDoNotKeep != null)
+                        foreach (var reference in compilationDoNotKeep.References)
+                            compilationDoNotKeep.GetAssemblyOrModuleSymbol(reference);
```

Comment:
all we're doing is simply walking the public surface area of the compilation.  This should not cause any differences in behavior.